### DISCUSSION TITLE
Skill tree tabs within specialization tabs (tabs with tabs)?

### DISCUSSION THREAD

**Discussion by WithinAmnesia** [flareteam/flare-game#888] [Q&A] [answered]
Is it possible to have the skill tree tabs (Wizard, Warrior etc.) in Flare have their own specialization tabs? Such as Wizard (tab) with 9 specialization tabs (Light, Dark, Order, Chaos etc. tabs)? I am having a very hard time fitting 9 specializations into one (enlarged) tab page and I am forced to make straight lines of skill specializations / abilities. Is it possible with the current version of Flare to have a skill tree tab page within a skill tree tab? Kind of like this in the flow of actions: Skill tree (open) -> Wizard (tab) -> Light (tab) [with light magic abilities].

**Comment by dorkster** [comment]
This would clutter the UI too much in my opinion. If anything, I'd much rather improve the behavior of the existing tabs when they don't all fit in the menu (e.g. if they have long titles or if there are too many tabs).

**Reply by WithinAmnesia** [reply]
Is it possible though or is it not doable in the current engine? I am trying to figure out how to fit 30 half-screens (~15 screens) worth of abilities into the skill tree and not have a page system where one build is 'first on the stack' of ~'pages'. I have 9 elemental types + physical and there are 5 build / play style trees with at least 9 elements available in said tree. I am running out of space with one tab per gameplay type. Each 'tree' (e.g. Wizard Light) has 100 levels or about ~20 'rungs' / ability levels (one 'tier' / 'rung' per ~5 levels). So I end up with a skilldrassil / Path of Exile skill tree https://poeplanner.com/ but I have no room to implement it in the current system. My tabs are maxed out for screen real-estate, in those tabs there is not enough room even with the entire screen full without micro buttons but then that doesn't work with smaller screens / it is bad design to use a magnifying glass to see spell abilities. I am wondering how to get more screen real-estate on the skill tree menu without making a book and I am just not sure how to do it without maybe another level of tabs. It is like each tab is a class with 9 specializations or trying to fit this times 3 into one skill tree https://classic.wowhead.com/talent-calc .

**Comment by WithinAmnesia** [comment]
Also is it currently possible in the Flare engine to unlock / 'learn' new possible abilities in the skill tree via an item like a tome / book? Like have hidden abilities that can be 'discovered' and are not learned (when the 'ability tome' is read) but possible to invest points into? Can the possible skill tree be mechanically expanded with tomes / items / skill trailers / a sort of event? I know the ability at the end of the Empyrean campaign but I wonder if a skill not active but still shows up in the skill tree for possible invest choice can be possible?

**Reply by dorkster** [reply]
This should be possible. The "Sacrifice" ability you're talking about is only enabled once a certain campaign status is set. So you could have a tome set a status to unlock the power. For upgrades, you could have them unlocked with points or other tomes. But if an upgrade requires a point, you can't "spend" a tome in place of the point.

**Reply by WithinAmnesia** [reply]
Okay that sound just like what I was hoping for. So a ~'tome' / event can 'unlock' a 'hidden' skill tree abilities and they can be invested into with skill points to activate; but one cannot use a 'tome' to activate said 'unlocked' / 'hidden' skill point? If that is the case that would be great to use for say skill trainers or little bits of flavor and or places around the world can offer unique skill rewards / 'skill unlocks'.

I remember in Baldur's Gate / Dungeons and Dragons that a Mage class could really only get new spells by learning them from a spell scroll / tome. Same with Diablo 1 I think (minus starting abilities). Although most games (like Flare's Empyrean Campaign too) the player can see all of the potential abilities. I am hoping that I can do a hybrid where there is a 'core branch' in a play style skill tree 'tab'. Say like a 'Wizard' with 'Light' specialization would have a 'core branch' of almost 'regular' player expectations of abilities, like 'Lightbolt' -> 'Lightball' -> 'Lightblast' etc. Yet there could also be like a unique place / specialized places in the world that could say have a ~'Tome of Lightshield' or something like that. So that there is a 'core branch' for progression but still that sense of mystery and discovery / diversity for a player to explore and enjoy :-).

Hopefully a sort of positive combination of both ways of gaining abilities, a 'core branch' from the stable way of a predefined path and a 'sense of mystery, discovery and adventure' from a hopefully exciting mystery wondering hunting path for potentially realizing new unknown possibilities.

**Comment by WithinAmnesia** [comment]
"This would clutter the UI too much in my opinion. If anything, I'd much rather improve the behavior of the existing tabs when they don't all fit in the menu (e.g. if they have long titles or if there are too many tabs)."- Dorkster

I don't want to change Flare empyrean campaign skill tree, I hope that was not inferred XD. I am making a bunch of pizzas so sorry if my comments are a bit rushed / sporadic in nature. I got so many abilities and not enough menu space lol. Is there a way to get more menu space that I am just not figuring out? I only know of an incomplete collection of ways to tweak the engine / mod Flare. How would you fit *let's see* 5 (play styles) times 9 (specializations of energy at least with just 9 elements) times 20 (level rungs) times 3 (at least for some would have multi skill-point investment options) equals at least 2,700 abilities. How do I put 2,700 abilities into flare's skill tree? I think vanilla WoW has https://classic.wowhead.com/spells/abilities (1,335 spells) + https://classic.wowhead.com/spells/talents (1,424 spells) = 2,759 (spooky) spells / abilities. So I think I need to find a way to get more screen real estate somehow for my ambitious omni class specialization system.

How can this be done within Flare's skill tree?

**Reply by dorkster** [reply]
Currently, the only way to get more real estate for powers is to add more tabs.

One thing that might make the situation better would be if the contents of the tabs could be scrolled (like the quest log). I think that would be a reasonable solution for the majority of games (maybe not ones with 2700 spells).

**Reply by WithinAmnesia** [reply]
Oh yeah that is a great idea! See simple things like that one can miss lol. So you think an extra layer of tabs within a tab and scrolling could do the job for ~2,700 abilities? I know it sounds like a lot but some spells are almost like recoloured duplicates or tweaks of a 'template spell' like ~'firebolt', is like ~'Icebolt' is like ~'darkbolt' etc. times 9. Although it is doable from a game dev stand point (just make one set of 20 ability tiers / rungs for each element to start for the 5 play styles melee weapons, bows, spirit, magic, hybrid). Yet from a skill tree stand point it would have ~2,700 abilities at once eventually. It is not completely bonkers it is just takes time to cover all of the play style and element combinations and give each one a unique twist / flavour / niche / utility / role. Although I am trying to future proof as much as I can so that I can have enough room to grow the skill tree out within reason / have room for future growth / be future proof responsible XD.

**Comment by WithinAmnesia** [comment]
Here is an image concept of a skill tree with 5 gameplay archetype tabs each with 9 elemental specialization tabs (5 x 9 tabs = 45 tabs):
![Possible Ability Skilldrassil](https://user-images.githubusercontent.com/86871191/150628712-82cd7038-40fe-4a6e-9d32-f6a48010203a.png)
Is this possible / doable with the Flare engine?
[Also a scrolling tab could fit the 20 ability tiers / 'rungs' for I think currently ~14-15 ability tiers / 'rungs' are 'possible' if they are all crammed in there; scrolling would be better XD.]

**Reply by dorkster** [reply]
Still going to say no to the sub-tabs. If the tab contents were scroll-able, the contents of those 9 sub-tabs could be stacked vertically inside the scroll area. The scroll area background image could be used to group the different sets of powers visually. Maybe a little tougher to navigate with such a large volume of powers, but at least everything could fit.

Opinion: I think the sort of skill tree design you're going for is an "acquired taste" both from a player's perspective and a game developer's perspective. I'd call it "bloat". It's not a design style I particularly want to encourage with Flare.

**Reply by WithinAmnesia** [reply]
Short-form: Can up and down and left and right scroll bars be used in a Flare menu to enable a 40 x 40 visual ability icon square grid without going crazy on tabs? The two scroll bars are better than ~infinite tabs. I am not sure how complex a menu zoom in and out is to implement so I am hesitant to ask for a zoom to be implemented. Two scroll bars should be superior to a bunch of tabs and it brings my video game in Flare closer to my original body of work of ~15+ years; so two birds with one stone :-).
~--------------
Long-form / 'discovering' a potentially better idea (circle + two scroll bars; maybe zoom). If I have the entire screen a skill tree I could use a scroll to increase the depth and there should be enough space. You are right with the bloat. I have to try to get the play style diversity without the unnecessary visual complexity. Simplicity is the ultimate form of sophistication. Also the bottom of the skill tree / inventory / menu overlaps with the bottom H.U.D. menu buttons; this it needs to be shortened to not conflict / overlap regardless. Also with scrolling and a wider area all of these play style combinations options can be found in the same area. I would like tabs within tabs for it offers a kind of depth / freedom of player exploration of play-styles that I am trying to realize. Ideally in my heart of hearts I would like no tabs and have a huge circle like Path of Exile so the mission is trying to fit that kind of depth into Flare.

I should be fluid to adapt yet focused in trying to make a it happen... Wait. Can the scroll bars work in both up and down? What if I don't even need tabs? I am not sure how complex zooming in and out of menu is yet if a menu can scroll to the left and right, up and down then I could have one 'circle' and be closer to my original hopes in the first place.

Could a more condensed version of this be in a giant scrolling menu (scrolls up and down, left and right, maybe zoom in and out as an extra(not sure if zooming is complex): https://poeplanner.com/ . Wait. How many screen icons do I really need? If I have say an ability have ~10 ranks and with ~20 different spell types that are found in each element 'small wing / branch'. I could have 5 play-style archetype 'big wings / branches' and in each of those 'big wings / branches' would have ~20 icons (with multiple / ~10 ranks), So 5 x 20 ... Wait. https://www.deviantart.com/withinamnesia/art/The-Jewelled-Eye-Of-AGABAL-Jeweyel-883211893 I could make this if I have enough screen real estate. Each colour could be an element, in each element 5 play style archetypes. So ~20 visual icons per one of the five play-style archetype is 20 x 5 =100. Then multiply that by 8 outer rings = 800 visual icons. Then some other system for the center, ~10 per shade of grey times 5 play style archetypes = 50 prey shade of grey, times 7 shades of grey = 350 icons. 350 + 800 = 1,150; damn that is a lot hmm, what ever maybe I can have a better idea later for less icon bloat. How many icons do I need without gutting the play style options? I said ~2,700 total abilities before and each one has 10 points of investment then I could that down to ~270 visual icons. (~270 + ~1,150) / 2 = ~710 visual icons average from the most to the least on a rough estimate. https://savagebirdlearning.com/2016/11/06/circles-squares-pi-78-5/ ~78.55% of a square is a circle so with about a ~10% loss or borders in order to fit ~1,000 visual icons (better to have too many than not enough) on one menu in a circle would require a grid square of (78.55 / 100 = 1.2730744748567791215786123488224, is need to add but to the 78.55 / 100.00 loss to get back to 100%), So a circle of 1,000 icons in a grid could roughly need (1.2730744748567791215786123488224 times ~1,000 = ~1,273.0744748567791215786123488224 squares, with a ~10% loss around the borders that is 1,273.0744748567791215786123488224 / 0.9 = 1,414.5271942853101350873470542471 grid squares total to reasonably fit ~1,000 visual icons into a gird square. 1,414.527 / ~1,415  grid squares in a square shape has the dimensions of 1,414.5271942853101350873470542471 square root 2  = 37.61020066797450809750285276529. Or to fit ~1,000 icons in a square grid (accounting for a ~10% border loss) would need a grid approximately ~37.61020066797450809750285276529 or (~38 x ~38) in size. So with extra margin a ~40 x 40 square would put some 'meat on the bone' / establish a future proof buffer. Okay time to simply that lol; but I learned something! If I can ~somehow get a ~40 x 40 grid of ability icons I should be set to make my crazy skilldrassil on one menu. I hope that two scroll bars could do that. Wait do I even need a side to side scroll bar? How many icons fit across the screen? *Estimates* about ~20 something it looks like. So yeah two scroll bars up and down and side to side (with maybe a zoom if it is not a hassle possible) then that should fulfill the crazy skilldrassil skill specialization tree.

**Reply by dorkster** [reply]
Zooming is out of the question as far as I'm concerned.

But horizontal scrolling should work in theory. WidgetScrollBox (and WidgetScrollBar) only has vertical scrolling support at the moment, so support for that would need to be added.

**Reply by WithinAmnesia** [reply]
Okay zooming is off the table all together then XD. So if the vertical scrolling is implemented in the skill tree, is horizontal scrolling on the lesser complex side of difficulty to implement? Is horizontal scrolling a switch of an axis or is there more to it than that (my guess would be yes yet I'm not sure how much more)?

**Reply by dorkster** [reply]
I don't think it will be too complex:

- Support creating a horizontal scrollbar in WidgetScrollBar
- Support defining the width of WidgetScrollBox contents in addition to height
- Add a horizontal WidgetScrollBar to WidgetScrollBox
- Ensure that gamepad navigation works. Left/Right is currently used to "exit" the WidgetScrollBox contents an move to a neighboring widget. This behavior would need to only happen at the left/right edges when navigating the contents.

**Comment by WithinAmnesia** [comment, marked as answer]
Okay so it looks like it is possible within reason to deeply expanded the Flare Engine skill tree via the use of vertical and horizontal scroll bars for the skill tree menu. Of which the vertical scroll bars (I think at least) need some form of engine work to implement / they are a derivative of the vertical scroll bar system / mechanic. I'll add an issue of horizontal menu scroll bars in the engine issue page just to 'keep tabs' (puns XD) on this issue and relax everyone with not having to actively remember about it, the issue can just sit passively like a sticky note hopefully.
